src/data_source/fetching.rs

```diff
@@ -563,19 +660,21 @@ where
     ///   leaf by height but not by hash, since we can't guarantee that a leaf with an arbitrary
     ///   hash exists.
     ///
-    /// If we do not trigger an active fetch for an object, but the object does in fact exist, we
-    /// will still eventually receive it passively, since we will eventually receive all blocks and
-    /// leaves that are ever produced.
+    /// If we do trigger an active fetch for an object, the provided callback will be called if and
+    /// when the fetch completes successfully. The callback should be responsible for notifying any
+    /// passive listeners that the object has been retrieved. If we do not trigger an active fetch
+    /// for an object, this function does nothing.
+    ///
+    /// In either case, as long as the requested object does in fact exist, we will eventually
+    /// receive it passively, since we will eventually receive all blocks and leaves that are ever
+    /// produced. Active fetching merely helps us receive certain objects sooner.
     async fn active_fetch<S, P>(
```

Comment:
Biggest change in all of this is how active fetching works. Basically because active fetching happens in a separate task, and may not happen at all depending on a decision taken after the task is already spawned, it was a real pain to try to have active fetch return the fetched object. It's not really necessary either, since we eventually get the object via the `Notifier`. It ended up being a lot simpler, IMO, to just use callbacks as needed to post-process fetched objects